question.php

```diff
@@ -99,15 +100,33 @@ private function update_ui(attempt_ui $ui): void {
      *      being started. Can be used to store state.
      * @param int $variant which variant of this question to start. Will be between
      *      1 and {@see get_num_variants()} inclusive.
-     * @throws moodle_exception
+     * @throws Throwable
      */
     public function start_attempt(question_attempt_step $step, $variant): void {
-        $attempt = $this->api->package($this->packagehash, $this->packagefile)->start_attempt($this->questionstate, $variant);
+        global $PAGE, $USER;
 
-        $this->attemptstate = $attempt->attemptstate;
-        $step->set_qt_var(self::QT_VAR_ATTEMPT_STATE, $attempt->attemptstate);
-        $this->scoringstate = null;
-        $this->update_ui($attempt->ui);
+        try {
+            $attempt = $this->api->package($this->packagehash, $this->packagefile)->start_attempt($this->questionstate, $variant);
+            $this->attemptstate = $attempt->attemptstate;
+            $step->set_qt_var(self::QT_VAR_ATTEMPT_STATE, $attempt->attemptstate);
+            $this->scoringstate = null;
+            $this->update_ui($attempt->ui);
+        } catch (Throwable $t) {
+            // Trigger server request error event.
+            // TODO: get question_attempt-id via DB-query (question_attempt_step-id -> question_attempt-id)?
+            $message = "The user with id '{$USER->id}' encountered an error in the question with id '{$this->id}' when starting " .
+                "a new attempt in the {$PAGE->context->get_context_name()} with id '{$PAGE->context->id}':\n{$t->getMessage()}";
```

Comment:
Du fügst hier die contextid ein, nicht die Quiz ID. Du musst den Kontext hier nicht extra beschreiben, da er in der Liste von Events ja sowieso nochmal verlinkt wird (in der Spalte "Event context").

Mein Wunsch, dass in der Nachricht die Quiz-ID genannt wird, war aber vielleicht etwas kurzsichtig. Moodle-Fragen können ja auch in anderen Aktivitäten verwendet werden. Wahrscheinlich reicht eine Nachricht wie

> The user with id '???' encountered an error when starting a new attempt of the question with id '???':

Vielleicht könnte man noch die question attempt ID nennen und den Namen des Fragetyps.

Reply:
> Du fügst hier die contextid ein, nicht die Quiz ID. Du musst den Kontext hier nicht extra beschreiben, da er in der Liste von Events ja sowieso nochmal verlinkt wird (in der Spalte "Event context").

Gleiches gilt auch für die User-ID, ich hatte das ursprünglich für das `debugging()` drin.

Zum letzten Punkt habe ich erstmal nur ein TODO geschrieben.

Reply:
Stimmt eigentlich, dass für die User-ID dasselbe gilt. Allerdings wird die User-ID bei den meisten Events, die mir bisher so untergekommen sind, in der description nochmal extra angegeben. Die Formulierung beginnt meist mit "The user with the id '$this->userid' ..." (siehe das Beispiel in der [Doku](https://docs.moodle.org/dev/Events_API#Methods)).